### PR TITLE
Add fine grained control on library tools builds

### PR DESCRIPTION
**Pull request by TheNicker**

### CODE REVIEW

**Comment by dcommander** [comment]
What is the purpose of this?

**Comment by TheNicker** [comment]
> What is the purpose of this?

Saves compilation time
Cleaner binary output dir

If I don't need those I should be able to opt these out from the build.

Just need the turbo-jpeg library

**Comment by dcommander** [comment]
That's why compile targets exist. They allow you to build only the items you need, as opposed to opting certain items out of the "all" target.

**Comment by dcommander** [comment]
Also please submit an issue to discuss such matters in the future rather than a PR. We almost never accept unsolicited PRs.

**Comment by TheNicker** [comment]
> That's why compile targets exist. They allow you to build only the items you need, as opposed to opting certain items out of the "all" target.

By compile targets, do you mean running cmake with - - target , or something else? Because the forner doesn't really solve my use case

**Comment by dcommander** [comment]
No, I mean doing something like `make jpeg` to build only the libjpeg API library.

Our build system behavior is consistent with that of other projects. We cannot add customizations for everyone's personal preferences, or our project would turn into a crufty unmaintainable mess. Patches must have demonstrable benefits to a broad range of users.

**Comment by TheNicker** [comment]
@dcommander ,
I completely agree that a library should solve the general use case, and no single entity should enforce it's custom preference.

However opt in/out libraries that are no mandatory are far from "custom preference".
There are many libraries that do exactlly as presented in this PR
E.g. in my project several image codec libraries are used. In WebP, which is considered pretty  much industry standard the following is applied before including the library:

```
#override webp default settings.
option(WEBP_BUILD_ANIM_UTILS "Build animation utilities." OFF)
option(WEBP_BUILD_CWEBP "Build the cwebp command line tool." OFF)
option(WEBP_BUILD_DWEBP "Build the dwebp command line tool." OFF)
option(WEBP_BUILD_GIF2WEBP "Build the gif2webp conversion tool." OFF)
option(WEBP_BUILD_IMG2WEBP "Build the img2webp animation tool." OFF)
option(WEBP_BUILD_VWEBP "Build the vwebp viewer tool." OFF)
option(WEBP_BUILD_WEBPINFO "Build the webpinfo command line tool." OFF)
option(WEBP_BUILD_LIBWEBPMUX "Build the libwebpmux library." OFF)
option(WEBP_BUILD_WEBPMUX "Build the webpmux command line tool." OFF)
option(WEBP_BUILD_EXTRAS "Build extras." OFF)
option(WEBP_BUILD_WEBP_JS "Emscripten build of webp.js." OFF)
```
11 Targets has been opted-out

Moreover the library currently overruling the **zero overhead principal of C++**. I do pay for what I don't use.
that relates to  before:

> But I would go further and suggest to deafult for a minimal build.
> And when using the CI/CD system opt in full build / benchmarks and tests

**Comment by dcommander** [comment]
> There are many libraries that do exactlly as presented in this PR

You presented one example. One is not "many." Regardless, I doubt that the WebP options in question are as ridiculously fine-grained as the ones you propose for libjpeg-turbo.

> Moreover [libjpeg-turbo](https://github.com/libjpeg-turbo/libjpeg-turbo) currently overruling the **zero overhead principal of C++**. I do pay for what I don't use.

I have no idea what you mean by that, but you seem to be raising a stink about a few seconds of extra build time.

> > But I would go further and suggest to deafult for a minimal build.
> 
> > And when using the CI/CD system opt in full build / benchmarks and tests

No way. More users expect to be able to run `make` and then `make test` than expect what you expect. If I did what you suggest, then I would have to field numerous support requests from those users, whereas you are the only person in the 13-year history of this project who has asked for what you're asking. The targets you want to exclude are needed by `make test`, and there is no way to make those targets contingent upon the `test` target. Furthermore, it is pretty ridiculous to introduce a separate CMake variable to control each target. That violates the KISS principal, because there is already another way to accomplish the same thing. If you are only using the libjpeg API, then you can already disable the build of TJUnitTest and TJBench by setting `WITH_TURBOJPEG=0`. Regardless, what you are asking for is a method to disable your ability to test your build of libjpeg-turbo. I do not find that to be a good idea, and I do not find your argument in favor of it to be compelling. If you are using `ExternalProject_Add()`, which is the only sanctioned method of including libjpeg-turbo as a subdirectory of another project, then you can specify the build command, which allows you to build only the targets you need. If you aren’t using `ExternalProject_Add()`, then you are outside of our support scope.

In other words, I don't buy your argument that this is generally useful at all.
